# Chapter 13 Concurrency 

## Introducing Concurrency

Concurrency is whenever things are happening "at the same time" or in parallel. 

JS is a single-threaded language, and it can only execute one task at a time.

The quick switching between tasks can give the illusion of concurrency, even though the single-thread characteristic.
- e.g. Ask the browser to get the user's geographical location while continue console log messages.

### Example 13-1: Get the user's geographical location

```javascript
console.log('Start');
// get the user's geographical location
navigator.geolocation.getCurrentPosition(
  position => console.log(position),   // callback function invoked by the Web API
  err => console.log(err)
);
console.log('Finish');
```
When you run the above code in the browser, you will see the following output:

```
Start
Finish
// the user's geographical location or an error message
```

![](img/24-08-06-10-50-50.png)

## JS engine and Browser

An interesting question arises from the above example:
- Who executes the task of getting the user's geographical location?
- Is it the JS engine or the browser? The browser is. 

The browser is implemented to have the multi-threaded capability.

So, the important concept to keep in mind: JS engine and the browser are two different runtime environments.

<figure>

![](img/24-08-06-11-04-36.png)

<caption>JS engine and the browser are two different runtime environments.

Fig Source: https://www.youtube.com/watch?v=eiC58R16hb8

</caption>

</figure>



## Asynchronous Programming

In JavaScript, concurrency is achieved through the use of asynchronous programming.

There are three ways to write asynchronous code in JavaScript:
- Callbacks
- Promises
- Async/Await

## Callbacks 

Callbacks are functions passed as arguments to other functions.
- They are invoked by other functions. 

An typical example of using callbacks is the `setTimeout` function.
- the `setTimeout` function is a Web API provided by the browser.

```javascript
console.log('Hi there!'); 
function greeting(waitTimeSeconds) {
    console.log(`Sorry for the wait ${waitTimeSeconds} seconds.`)
}
setTimeout(greeting, 2000, 2); // async call; Execute the greeting function after 2 seconds
console.log('Please wait'); 
```

The outputs of the above code are:
```
Hi there!
Please wait
Sorry for the wait 2 seconds.
```

### Callback Hell

The disadvantage of using callbacks is the callback hell.
The nested callbacks make the code hard to read and maintain.

### Example 13-2 : Write a function that logs messages every second for four seconds.

```javascript
// ex_13_02.js
function startTimeouts() {
    setTimeout(() => { // first callback
        console.log('First timeout');
        setTimeout(() => { // second callback
            console.log('Second timeout');
            setTimeout(() => { // third callback
                console.log('Third timeout');
                setTimeout(() => { // fourth callback
                    console.log('Fourth timeout');
                    // Continue nesting if needed
                }, 1000);
            }, 1000);
        }, 1000);
    }, 1000);
}

// Call the function
startTimeouts();
```

The output of the above code is:
```
First timeout
Second timeout
Third timeout
Fourth timeout
```

To avoid the callback hell, we can use Promises or Async/Await (covered later).

## Lab 13-1: Add an event listener (function) to a button element 

TBD 

## Promises

An async function returns a Promise object when it completes its task.
- (or the async function resolves the task and return the promise object as the result.)
- Then, the Promise object is placed in a special place called the Microtask Queue in JS engine. 
- The JS engine will find a suitable time to execute the callbacks associated with the promise object.


## Use the Promise object 

A Promise object has three states:
- Pending: The initial state of the promise object.
- Fulfilled: The promise object is resolved. (The async task is completed successfully.)
- Rejected: The promise object is rejected. (The async task is failed.)


Terms to know:
- Resolved promise: The promise is settled or "locked-in" to match the eventual state of another promise, and further resolving or rejecting it has no effect.
  - A Promise object can only be resolved once.
- Settled promise: A promise that is either resolved or rejected, but not pending.

<figure

![](img/24-08-06-23-33-44.png)

<caption>Promise States <br/>

Source: [Promise - JavaScript | MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#constructor)

 </caption>

</figure>

## Register handlers for the Promise object


We must assign the handlers (or callback functions) to the promise object to handle the fulfilled and rejected states respectively.

Use the [`then` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) of the Promise object to register the handler for the fulfilled state.

Use the [`catch` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch) of the Promise object to register the handler for the rejected state.





### Example 13-3 : Use `fetch()` web API to get the data from a URL.

`fetch()` return a Promise object. We can register the handles using the `then` and `catch` methods.

```javascript
fetch(url)
        .then(response => {  // register a callback to handle the response
            console.log("First handler for the fetch()'s promise")
            if (!response.ok) {
                throw new Error('Network response was not ok');
            }
            // response.text() returns a promise that resolves with the response body
            return response.text();
        })
        .catch(error => {
            console.error('Error fetching data:', error);
        });
```

The first arrow function is the handler for the fulfilled state.
- It will be invoked when the promise object is resolved.
- If the response status is ok, the handler will return the response body as a text.

The second arrow function is the handler for the rejected state.
- It will be invoked when the promise object is rejected.
- It will log the error message to the console.

Next, let us wrap the above code in a function and call it.

```javascript
// A function that returns a promise object
function fetchData(url) {
    return fetch(url)
        .then(response => {  // register a callback to handle the response
            console.log("First handler for the fetch()'s promise")
            if (!response.ok) {
                throw new Error('Network response was not ok');
            }
            // read the response body as json
            // response.json() returns a promise that resolves with the response body
            return response.json();
        })
        .catch(error => {
            console.error('Error in the first promise:', error);
        });
}
```

when you want to print the response body, you can call the `fetchData` function as follows:

```javascript
const url = "https://jsonplaceholder.typicode.com/posts/1";
fetchData(url)
    .then(data => console.log(data))
    .catch(error => console.error(error));
``` 

The above code will print the response body of the URL to the console.

```
Fetching data from https://jsonplaceholder.typicode.com/posts/1
Please wait for the data to be fetched...
First handler for the fetch()'s promise
Second handler for the response.json()'s promise
Data fetched: {
  userId: 1,
  id: 1,
  title: 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
  body: 'quia et suscipit\n' +
    'suscipit recusandae consequuntur expedita et cum\n' +
    'reprehenderit molestiae ut ut quas totam\n' +
    'nostrum rerum est autem sunt rem eveniet architecto'
}
```


see [ex_13_03.js](ex_13_03.js) for the complete code.

## Lab 13-2: Fetch data from a URL and show the first N characters of the response body

1. Write a function `displayData` that log the first N characters of the given string to the console.
2. Use `fetch()` to get the data from the URL "https://jsonplaceholder.typicode.com/posts/2". 
   - Register the `displayData` function as the handler for the fulfilled state.
   - Register an arrow function to log the error message to the console as the handler for the rejected state.

## Chain Promises

Promise objects are thenable objects 
- Each Promise object has a `then` method that returns a new Promise object.
- The `then` method can be chained to handle the resolved state of the promise object.
- You can treat the chained `then` method as a sequence of operations on returned Promise objects.

<figure>

![](img/24-08-06-22-53-08.png)

<caption>Chaining Promises <br/>

Figure Source: [JavaScript Promise Chaining â€” Basics](https://levelup.gitconnected.com/javascript-promise-chaining-basics-e2618c5e74a9)

</caption>

</figure>

### Example 13-4: Rewrite Example 13-3 using Promise chaining

Rewrite the Example 13-3 by using Promise chaining so that we don't need the `fetchData` function. 
Use the `fetch()` function directly.

```javascript
fetch(url)
    .then(response => {  // register a callback to handle the response
        console.log("First handler for the fetch()'s promise")
        if (!response.ok) {
            throw new Error('Network response was not ok');
        }
        // response.text() returns a promise that resolves with the response body
        return response.json();
    })
    .then(data => {
        console.log("Second handler for the response.json()'s promise");
        // data is the response body
        console.log("Second handler for the response.json()'s promise");
        console.log('Data fetched:', data);
    })
    .catch(error => {
        console.error('Error in the first promise:', error);
    });

console.log('Please wait for the data to be fetched...');
```


## Create your own Promise object

You might want to create and return your own Promise object to handle asynchronous tasks in your application, not just the Web APIs.

Use the `Promise` object to wrap the callback functions that do not support the Promise object (or not support the concurrency).

Steps to create a Promise object:
1. Create a new Promise object by calling the `Promise` constructor.
2. Pass a function as the argument to the `Promise` constructor, either an arrow function or a named function.
3. The function is called the executor function and it has two parameters: `resolveFunc` and `rejectFunc` parameters.
   - `resolveFunc` is a function that resolves the promise object.
   - `rejectFunc` is a function that rejects the promise object.

```javascript
function executor(resolveFunc, rejectFunc) {
  // code to execute the async task
}
```

4. You write the custom code in the executor function [1]. 
   - Use the `resolveFunc(value)` to resolve the promise object and pass the value to the handler.
     - the `value` can be another promise object.
     - The Promise object will change to the fulfilled state after the `resolveFunc` is called.
   - Use the `rejectFunc(reason)` to reject the promise object and pass the reason to the handler.
     - the `reason` usually is an error object.
     - The Promise object will change to the **rejected** state after the `rejectFunc` is called.
   - **Important**: If neither `resolveFunc` nor `rejectFunc` is called, the promise object will stay in the **pending** state.
   

### Example 13-5: Make a function do the async task and return a Promise object

Assume we have a function that take lots of time to complete the task:


```javascript
/**
 *  This function is a long running task that takes a while to complete.
 * @returns 
 */
function longtimeTask() {
  console.log('== Enter the long task');
  let i = 0;
  while (i < 1000000000) {
    i++;
  }
  console.log('== Exit the long task');
  // Return a random number between 0 and 100
  return Math.random() * 100;
}


console.log('Start the long running task');

// Use the Promise object to run the long running task
// When the promise is created, the long running task is started
const longtimeTaskPromise = new Promise((resolve, reject) => {
    const result = longtimeTask();
    if (result < 50) {
        reject('The result is less than 50');
    }
    resolve(result); // fulfilled 
});

// get the result of the long running task
longtimeTaskPromise
    .then(result => console.log('The result is', result))
    .catch(error => console.error('An error occurred:', error));

console.log('Please wait for the long running task to complete...');
```

The output of the above code is:

```
Start the long running task
== Enter the long task
== Exit the long task
Please wait for the long running task to complete...
The result is 61.054982580283635
```

<!-- readFile() using the Promise object (assignment) -->

You may interest in the following article about Promises:
[JavaScript Visualized: Promises & Async/Await](https://medium.com/@lydiahallie/javascript-visualized-promises-async-await-a3f1aad8a943)


Of course, you can create an async function by returning a Promise object.

The pattern of the function that returns a Promise object is as follows:

```javascript
function myAsyncFunction() {
    return new Promise((resolve, reject) => {
        // code to execute the async task
        // call resolve(value) to resolve the promise object
        // call reject(reason) to reject the promise object
    });
}
```

### Example 13-6: Rewrite the `longtimeTask` function to return a Promise object.

This example rewrites the `longtimeTask` function in Example 13-5 to return a Promise object.

```javascript
// ex_13_06.js
function longtimeTask(){
    return new Promise((resolve, reject) => {
        console.log('== Enter the long task');
        let i = 0;
        while (i < 1000000000) { // still run the task in JS engine, not Web API. 
            // NOT async operation
            i++;
        }
        // Return a random number between 0 and 100
        console.log('== Exit the long task');
        const result = Math.random() * 100;
        if (result < 50) {
            reject('The result is less than 50');
        }
        resolve(result); // fulfilled 
    });
}

// call the longtimeTask function
console.log('Start the long running task');

longtimeTask()
    .then(result => console.log('The result is', result))
    .catch(error => console.error('An error occurred:', error));

console.log('Please wait for the long running task to complete...');    
```

**Misconception** in the above code:
- The `longtimeTask` function does not run asynchronously, although it returns a Promise object.
- It run the LOOP in the JS engine, not in the Web API. 
- Because the JS engine is single-threaded, the loop will block the execution of other tasks in the JS engine.
- To run the task asynchronously, you need to offload the task to the Web API.
- Replace the `while` loop with the `setTimeout` function (Web API) to run the task asynchronously.
- See [ex_13_06_1.js](./ex_13_06_1.js) for the complete code.

## Async and Await

Use `then()` and `catch()` to specify the handlers (callbacks) for the Promise object.
That might cause the callback hell if you have many nested callbacks and make the code hard to read and maintain.

Use the async/await syntax to write asynchronous code that looks like synchronous code.

### Example 13-7: Rewrite Example `ex_13_06_01.js` using the async/await syntax

```javascript
// call the longtimeTask function
console.log('Start the long running task');
// Use an Immediate Invoke Function Expression (IIFE) to call the function returning a promise
(async () => {
    let result = await longtimeTask();
    console.log('The result is: ', result);
})()
console.log('Please wait for the long running task to complete...');  
```

`await` means wait for the promise object to be resolved.
- the `result` variable will store the fulfillment value of the promise object if the promise object is resolved.
- If the promise object is rejected, the `await` will throw an error with the rejection value of the promise object.
  - You can use the `try...catch` block to handle the error.
- The `await` keyword can only be used inside an async function.

### Example 13-8: What's wrong with the following code?

```javascript
// call the longtimeTask function
console.log('Start the long running task');
// Use an Immediate Invoke Function Expression (IIFE) to call the function returning a promise
let result = await longtimeTask();
console.log('The result is', result);
console.log('Please wait for the long running task to complete...');    
```

The above code will throw a syntax error because the `await` keyword can only be used inside an async function.
- Inside the async function, you write code to handle the resolved value or exception of the promise object.

```
SyntaxError: await is only valid in async functions and the top level bodies of modules
```

To fix the error, you can define a named function to wrap the code and call the function, or use an IIFE to call the function.

### Example 13-9: Fix the code in Example 13-8 by a named async function

```javascript
console.log('Start the long running task');
// Define a named async function 
async function runLongTimeTask(){
    try{
        let result = await longtimeTask();
        console.log('The result is', result);
    } catch(error){
        console.error('An error occurred:', error);
    }
}
// call the longtimeTask function
runLongTimeTask();
runLongTimeTask();
runLongTimeTask();
runLongTimeTask();
console.log('Please wait for the long running task to complete...');    
```

Output:
```
Start the long running task
== Enter the long async task
== Enter the long async task
== Enter the long async task
== Enter the long async task
Please wait for the long running task to complete...
== Inside the long async task
== Exit the long task
The result is 54.74726285881244
== Inside the long async task
== Exit the long task
An error occurred: The result is less than 50
== Inside the long async task
== Exit the long task
An error occurred: The result is less than 50
== Inside the long async task
== Exit the long task
The result is 78.86292440819362
```

### Example 13-9: Fix the code in Example 13-8 by a named async function


In summary, why do we use the async/await syntax?
- The async/await syntax is a syntactic sugar for the Promise object.
  - JS engine unpacks the Promise object for you.
- It makes the code easier to read and maintain 
  - because it looks like synchronous code and avoid the "callback hell".
- Can use the `try...catch` block to handle the error.


### The return value of the async function

Async functions always return a Promise object [2].

If the return value of an async function is not explicitly a promise, it will be implicitly wrapped in a promise.

For example:

The code:

```javascript   
async function myAsyncFunction() {
    return 'Hello World';
}
```

is equivalent to:

```javascript
async function myAsyncFunction() {
    return Promise.resolve('Hello World');
}
```

To call the above async function without the `await` keyword, you can use the `then` method of the Promise object.

```javascript
myAsyncFunction().then(value => console.log(value));
```

See [ex_13_10.js](ex_13_10.js) for the complete code.


## JS Runtime Model and Event Loop

-  JavaScript is a single-threaded language
   -  JS can only execute one task at a time
-  A thread in this context means a path of execution

- To handle asynchronous tasks, JS employs a runtime model. 


- JS runtime model consists of the following components:
    -  Call Stack: LIFO, the code to be executed in the JS engine 
    -  Web APIs: APIs to interact with the browser. Browser and the JS engine are two different runtime environments.
    -  Callback Queue: A queue to store the callback functions invoked by the Web APIs when APIs complete their asynchronous tasks.
    -  Event Loop: A loop to check the Call Stack and the Callback Queue. 
       -  The loop run tasks in the Call Stack first.
       -  If and only if the Call Stack is empty, the Event Loop will move the tasks from the Callback Queue to the Call Stack.
    -  Microtask Queue
       -  A queue to store the promise-based tasks. 
       -  Microtasks have higher priority than the tasks in the Callback Queue.
  
<fig> 

![](img/24-08-06-09-55-57.png)

Fig source: [JavaScript Runtime Environment: Web API, Task Queue and Event Loop](https://slawinski.dev/blog/javascript-runtime-environment-web-api-task-queue-and-event-loop/)

<caption>JS runtime model</caption>

</fig>

### Example 13-10: Use JavaScript Visualizer 9000 to visualize the runtime process for the code in ex_13_06_1.js

[JS Visualizer 9000 - An interactive JavaScript runtime model visualizer.](https://www.jsv9000.app/)

![](img/24-08-14-14-36-24.png)

Note: The JS Visualizer 9000 cannot accept the async/await syntax. 

## Video and Interactive Tools to help you understand the JS runtime model

[Lydia Hallie, 2024. JavaScript Visualized - Event Loop, Web APIs, (Micro)task Queue](https://www.youtube.com/watch?v=eiC58R16hb8)

[JS Visualizer 9000 - An interactive JavaScript runtime model visualizer.](https://www.jsv9000.app/)


# References 

1. [Promise() constructor - JavaScript | MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise)

2. [async function - JavaScript | MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function)